packages/puppeteer-core/src/common/Input.ts

```diff
@@ -439,87 +512,134 @@ export class Mouse {
   }
 
   /**
-   * Dispatches a `mousemove` event.
+   * Moves the mouse to the given coordinate.
+   *
    * @param x - Horizontal position of the mouse.
    * @param y - Vertical position of the mouse.
-   * @param options - Optional object. If specified, the `steps` property
-   * sends intermediate `mousemove` events when set to `1` (default).
+   * @param options - Options to configure behavior.
    */
   async move(
     x: number,
     y: number,
-    options: {steps?: number} = {}
+    options: MouseMoveOptions = {}
   ): Promise<void> {
     const {steps = 1} = options;
-    const fromX = this.#x,
-      fromY = this.#y;
-    this.#x = x;
-    this.#y = y;
+    const points: Point[] = [];
     for (let i = 1; i <= steps; i++) {
+      points.push({
+        x: this.#position.x + (x - this.#position.x) * (i / steps),
+        y: this.#position.y + (y - this.#position.y) * (i / steps),
+      });
+    }
+    await Promise.all(
+      points.map(async point => {
+        const lastPosition = this.#position;
+        try {
+          this.#position = point;
+          await this.#client.send('Input.dispatchMouseEvent', {
+            type: 'mouseMoved',
+            ...this.#position,
+            modifiers: this.#keyboard._modifiers,
+            buttons: this.#buttons,
+            // This should always be 0 (i.e. 'left'). See
+            // https://w3c.github.io/uievents/#event-type-mousemove
+            button: MouseButton.Left,
+          });
+        } catch (error) {
+          this.#position = lastPosition;
```

Comment:
Let's say we have two move points: p1 & p2. After we start Promise.all, this.#position is equal p2. Now let's assume that that dispatchMouseEvent for p2 succeeds first and call for p1 is rejected. The this.#position is then incorrectly set to p1 whereas the mouse is at p2?

Reply:
Obsolete